PXI Level Error Handling

prhuft opened the pull request:
Errors occurring in device classes are now lumped into XMLError and HardwareError types (where sensible) so they can be handled at the PXI level with only as much granularity as is needed to decide what state the server should be in. Handling these errors at the PXI level has also been implemented, such that devices which have issues initializing will be removed from the measurement loop until reinitialized correctly. I have not yet implemented functionality to mark devices as in an error state, e.g. if the error occurs during a measurement, but this can be done easily. There are other conceivable types of exceptions, e.g. TimeoutError that have not been implemented here but could be. Timeouts are something we still need to deal with.

JuanBohorquez3 commented on analogin.py:
```diff
@@ -138,6 +142,18 @@ def init(self):
+                # end the task nicely
+                self.stop()
+                self.close()
+                msg = '\n AnalogInput task initialization failed'
```
Do we want to preserve original error message? eg write:
`msg = f"{e}\n AnalogInput task initialization failed"`
instead.

prhuft replied:
yes, that is preferable. I'll go through and add that in

c-poole replied:
I don't think it is necessary. The trace-back should make it clear what errors were raised to eventually obtain the hardware error so we don't need to manually do that job.

prhuft replied:
@c-poole that was my original intent for how this would function. sounds good, i'll leave it for now

JuanBohorquez3 commented on analogin.py:
```diff
@@ -152,28 +168,21 @@ def is_done(self) -> bool:
+                # end the task nicely
+                self.stop()
+                self.close()
+                msg = '\n AnalogInput check for task completion failed'
```
same comment on error message as above

JuanBohorquez3 commented on analogin.py:
```diff
@@ -215,4 +236,59 @@ def data_out(self) -> str:
+            try:
+                self.task.stop()
+                
+            except DaqError as e:
+                self.close()
+                msg = '\n AnalogInput failed to stop current task'
+                raise raise HardwareError(self, task, message=msg)
+
+            except DaqWarning as e:
+                self.logger.warning(str(e.message))
```
seems like this code block is repeated a few times in this file (except for the message). Could you create a function that runs the function you want and takes care of this error handling?

prhuft replied:
yes, probably. i do like the transparency of this though, so I think it's okay for now at least.

c-poole commented on analogin.py:
```diff
@@ -86,39 +89,40 @@ def load_xml(self, node):
-
+        
+            except (KeyError, ValueError):
+                self.logger.exceptions()
```
`self.logger.exception()` without the s

c-poole commented on analogin.py:
```diff
@@ -86,39 +89,40 @@ def load_xml(self, node):
-                        self.logger.error(f"Not a valid {child.tag} value {child.text} \n {e}")
-                        raise
+                    except KeyError: 
+                        raise KeyError(f"Not a valid {child.tag} value {child.text} \n {e}")
```
There is no e to be referenced here in the message as currently written.

prhuft replied:
oops.

c-poole commented on analogin.py:
```diff
@@ -138,6 +142,18 @@ def init(self):
+                self.stop()
+                self.close()
+                msg = '\n AnalogInput task initialization failed'
+                raise HardwareError(self, task, message=msg)
```
Should this be `self.task`?

prhuft replied:
no.

c-poole replied:
Where is the variable task defined then?

c-poole commented on analogin.py:
```diff
@@ -152,28 +168,21 @@ def is_done(self) -> bool:
+                self.stop()
+                self.close()
+                msg = '\n AnalogInput check for task completion failed'
+                raise HardwareError(self, task, message=msg)
```
Should this be `self.task`?

prhuft replied:
no.

c-poole commented on analogin.py:
```diff
@@ -186,9 +195,21 @@ def get_data(self):
+                self.stop()
+                self.close()
+                msg = '\n AnalogInput failed to read data from hardware'
+                raise raise HardwareError(self, task, message=msg)
```
double raise, also `self.task`?

prhuft replied:
double raise to be fixed. no.

c-poole commented on analogin.py:
```diff
@@ -215,4 +236,59 @@ def data_out(self) -> str:
                 
-            return self.data_string
\ No newline at end of file
+            return self.data_string
```
Is there a reason to assign the output of this method to a instance variable? It doesn't seem to be used anywhere else.

prhuft replied:
it isn't used elsewhere at the moment, but it could be a nice thing to have data associated with a device rather than merely spat out to the pxi measurement function in routine looping.

c-poole commented on analogin.py:
```diff
@@ -215,4 +236,59 @@ def data_out(self) -> str:
+                self.stop()
+                self.close()
+                msg = '\n AnalogInput failed to start task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`?

prhuft replied:
no.

c-poole commented on analogin.py:
```diff
@@ -215,4 +236,59 @@ def data_out(self) -> str:
+            except DaqError as e:
+                self.close()
+                msg = '\n AnalogInput failed to stop current task'
+                raise raise HardwareError(self, task, message=msg)
```
double raise, `self.task`?

prhuft replied:
yes. no.

c-poole commented on analogin.py:
```diff
@@ -215,4 +236,59 @@ def data_out(self) -> str:
+                
+            except DaqError as e:
+                msg = '\n AnalogInput failed to close current task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

prhuft replied:
no.

c-poole commented on analogin.py:
```diff
@@ -215,4 +236,59 @@ def data_out(self) -> str:
+                raise HardwareError(self, task, message=msg)
+
+            except DaqWarning as e:
+                self.logger.warning(str(e.message))
```
`DaqWarning` doesn't appear to have a `message` variable. My understanding of warnings is that they print messages to stdout without halting the program. Is it necessary for us to intercept those and use the logger instead? This comment applies to all such instances of this except block.

prhuft replied:
if what you are suggesting is warnings will be logged automatically without explicit interception and logging, then yes, i can remove these.

c-poole commented on analogout.py:
```diff
@@ -94,10 +102,10 @@ def load_xml(self, node):
 
                 elif child.tag == "minimum":
-                    self.minValue = float(child.text)
+                    self.minValue = str_to_float(child.text)
```
str_to_float was removed in a previous PR and doesn't appear to have been re-inserted. Is there a new reason to re-introduce this function that we didn't discuss in the meeting on the XMLLoader functions?

prhuft replied:
this was a merge mistake, per my comment above somewhere.

c-poole commented on analogout.py:
```diff
@@ -120,9 +128,8 @@ def load_xml(self, node):
-                        self.logger.error(f"Not a valid {child.tag} value {child.text} \n {e}")
-                        raise
+                    except KeyError:
+                        raise KeyError(f"Not a valid {child.tag} value {child.text} \n {e}")
```
There isn't an `e` to reference here as written.

c-poole commented on analogout.py:
```diff
@@ -131,10 +138,14 @@ def load_xml(self, node):
                 else:
                     self.logger.warning(f"Unrecognized XML tag \'{child.tag}\' in <AnalogOutput>")
+                    
+           except (KeyError, ValueError):
```
Indentation doesn't match corresponding try block

c-poole replied:
Still doesn't match.

c-poole commented on analogout.py:
```diff
@@ -131,10 +138,14 @@ def load_xml(self, node):
                     self.logger.warning(f"Unrecognized XML tag \'{child.tag}\' in <AnalogOutput>")
+                    
+           except (KeyError, ValueError):
+                self.logger.exceptions()
```
`logger.exception`

c-poole commented on analogout.py:
```diff
@@ -148,60 +159,90 @@ def init(self):
-                        sample_mode=AcquisitionType.FINITE, # default
-                        samps_per_chan=1000) # default
+                try:
+                    if self.task != None:
```
Comparison with None should be done using `is` and `is not` unless the object being compared has custom equality operators that produce a situation where `foo is None` is `False` but `foo == None` is `True`.

c-poole commented on analogout.py:
```diff
@@ -148,60 +159,90 @@ def init(self):
+                            self.stop()
+                            self.close()
+                            msg = '\n AnalogOutput failed to close current task'
+                            raise HardwareError(self, task, message=msg)
```
`self.task`

prhuft replied:
no.

c-poole commented on analogout.py:
```diff
@@ -148,60 +159,90 @@ def init(self):
+        if not (self.stop_connections or self.reset_connection) and self.enable:
+                        
+            channels, samples = self.waveforms.shape
+            sample_mode = AcquisitionType.FINITE 
```
`sample_mode` does not appear to be used.

c-poole commented on analogout.py:
```diff
@@ -148,60 +159,90 @@ def init(self):
+                self.stop()
+                self.close()
+                msg = '\n AnalogOutput hardware update failed'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on analogout.py:
```diff
@@ -215,8 +256,19 @@ def is_done(self) -> bool:
+                self.stop()
+                self.close()
+                msg = '\n AnalogOutput check for task completion failed'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on analogout.py:
```diff
@@ -227,13 +279,51 @@ def start(self):
+                self.stop()
+                self.close()
+                msg = '\n AnalogOutput failed to start task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on analogout.py:
```diff
@@ -227,13 +279,51 @@ def start(self):
+            except DaqError as e:
+                self.close()
+                msg = '\n AnalogOutput failed to stop current task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on analogout.py:
```diff
@@ -227,13 +279,51 @@ def start(self):
+                
+            except DaqError as e:
+                msg = '\n AnalogOutput failed to close current task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on analogout.py:
```diff
@@ -227,13 +279,51 @@ def start(self):
+        Close the task
+        """
+        
+        if self.task != None:
```
`self.task is not None`

prhuft commented on analogout.py:
```diff
@@ -94,10 +102,10 @@ def load_xml(self, node):
 
                 elif child.tag == "minimum":
-                    self.minValue = float(child.text)
+                    self.minValue = str_to_float(child.text)
```
str_to_float should be float. I made an error when merging master into my branch and i will fix this

c-poole commented on digitalin.py:
```diff
@@ -31,23 +36,26 @@ def __init__(self, pxi):
+        self.isInitialized = False
         
-            assert node.tag == self.expectedRoot
+        assert node.tag == self.expectedRoot, "Expected tag "+
```
Line continuation error missing the \. I would recommend changing to this:

```
        assert (node.tag == self.expectedRoot,
                f"Expected tag <{self.expectedRoot}>, but received <{node.tag}>")
```

c-poole commented on digitalin.py:
```diff
@@ -56,7 +64,11 @@ def load_xml(self, node):
-                    
+                            
+                except ValueError:
+                    self.logger.exceptions()
```
`self.logger.exception()`

c-poole commented on digitalin.py:
```diff
@@ -67,24 +79,48 @@ def init(self):
                         
             # Clear old task
             if self.task != None:
```
equality operator comparison with None

c-poole commented on digitalin.py:
```diff
@@ -67,24 +79,48 @@ def init(self):
+                    self.stop()
+                    self.close()
+                    msg = '\n TTLInput failed to close current task'
+                    raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalin.py:
```diff
@@ -67,24 +79,48 @@ def init(self):
+                #   task.read will be of type int
+                self.task.di_channels.add_di_chan(
+                    lines=self.lines
+                    name_to_assign_to_lines=u'', # this looks like a typo but came from nidaqmx docs...
```
Multiple syntax issues in this function call.

prhuft replied:
besides the <LineGrouping...> problem, which has been fixed, and the missing comma after self.lines, is there something else?

c-poole replied:
I think those are the only issues here.

c-poole commented on digitalin.py:
```diff
@@ -67,24 +79,48 @@ def init(self):
+                self.stop()
+                self.close()
+                msg = '\n TTLInput hardware initialization failed'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalin.py:
```diff
@@ -104,24 +140,34 @@ def check(self):
+                self.stop()
+                self.close()
+                msg = '\n TTLInput data check failed'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalin.py:
```diff
@@ -143,4 +189,57 @@ def data_out(self) -> str:
+                self.stop()
+                self.close()
+                msg = '\n TTLInput failed to start task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalin.py:
```diff
@@ -143,4 +189,57 @@ def data_out(self) -> str:
+                self.task.stop()
+            except DaqError as e:
+                msg = '\n TTLInput failed while attempting to stop current task'
+                raise HardwareError(self, task, message=msg)
```
`self.task1

c-poole commented on digitalin.py:
```diff
@@ -143,4 +189,57 @@ def data_out(self) -> str:
+                
+            except DaqError as e:
+                msg = '\n TTLInput failed to close current task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalout.py:
```diff
@@ -66,9 +76,8 @@ def load_xml(self, node):
-                                self.logger.error(f"Not a valid {child.tag} value {child.text} \n {e}")
-                                raise
+                            except KeyError:
+                                raise KeyError(f"Not a valid {child.tag} value {child.text} \n {e}")
```
missing `e`.

c-poole commented on digitalout.py:
```diff
@@ -85,47 +94,72 @@ def load_xml(self, node):
                     self.logger.warning(f"Unrecognized XML tag \'{child.tag}\' in <{self.expectedRoot}>")
+            
+            except (KeyError, ValueError):
+                self.logger.exceptions()
```
`self.logger.exception`

c-poole commented on digitalout.py:
```diff
@@ -85,47 +94,72 @@ def load_xml(self, node):
-             if self.enable:
-
                 # Clear old task
                 if self.task != None:
```
equality comparison with None

c-poole replied:
Also self.task not set equal to None in constructor so I'm not sure if this would work either way.

c-poole commented on digitalout.py:
```diff
@@ -85,47 +94,72 @@ def load_xml(self, node):
+                        self.stop()
+                        self.close()
+                        msg = '\n DAQmxDO failed to close current task'
+                        raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalout.py:
```diff
@@ -85,47 +94,72 @@ def load_xml(self, node):
+                    self.task.timing.cfg_samp_clk_timing(
+                        rate=self.clockRate, 
+                        active_edge=Edge.RISING, # default
+                        sample_mode=AcquisitionType.FINITE, # default
```
`AcquisitionType` not imported in this module

c-poole commented on digitalout.py:
```diff
@@ -85,47 +94,72 @@ def load_xml(self, node):
+                        rate=self.clockRate, 
+                        active_edge=Edge.RISING, # default
+                        sample_mode=AcquisitionType.FINITE, # default
+                        samps_per_chan=samplesPerChannel) 
```
`self.samplesPerChannel`? The bare variable doesn't appear to be defined anywhere in this function.

c-poole commented on digitalout.py:
```diff
@@ -85,47 +94,72 @@ def load_xml(self, node):
+                    # Write digital waveform 1 chan N samp
+                    self.task.write(
+                        self.data, 
+                        auto_start=AUTO_START_UNSET, #default
```
Where is the `AUTO_START_UNSET` constant defined?

prhuft replied:
this is what I get for pasting from documentation and not using an IDE that catches these for me

c-poole replied:
Using pycharm and scrolling through it yelling at me after I pulled the branch is how I generated 90% of this review. I suggest using pycharm.

c-poole commented on digitalout.py:
```diff
@@ -85,47 +94,72 @@ def load_xml(self, node):
+                    self.stop()
+                    self.close()
+                    msg = '\n DAQmxDO hardware initialization failed'
+                    raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalout.py:
```diff
@@ -134,15 +168,26 @@ def is_done(self) -> bool:
+                self.stop()
+                self.close()
+                msg = '\n DAQmxDO check for task completion failed'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalout.py:
```diff
@@ -153,7 +198,18 @@ def start(self):
+                self.stop()
+                self.close()
+                msg = '\n DAQmxDO failed to start task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalout.py:
```diff
@@ -162,4 +218,28 @@ def stop(self):
+                self.task.stop()
+            except DaqError as e:
+                msg = '\n DAQmxDO failed while attempting to stop current task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on digitalout.py:
```diff
@@ -162,4 +218,28 @@ def stop(self):
+        Close the task
+        """
+        
+        if self.task != None:
```
equality comparison with None

c-poole commented on digitalout.py:
```diff
@@ -162,4 +218,28 @@ def stop(self):
+                
+            except DaqError as e:
+                msg = '\n DAQmxDO failed to close current task'
+                raise HardwareError(self, task, message=msg)
```
`self.task`

c-poole commented on hamamatsu.py:
```diff
@@ -111,109 +112,119 @@ def load_xml(self, node: ET.Element):
+                    self.set_by_dict("trigger_polarity", child.text, self.TRIG_POLARITY_VALUES)
+
+                # this appears to not exist in the current master, but was in my branch.
+                # was it removed intentionally or should it be here?
```
I did not intentionally remove it if I did.

c-poole commented on hamamatsu.py:
```diff
@@ -363,7 +374,7 @@ def init(self):
             dtype=np.uint16
         )
-        self.is_initialized = True
+        self.isInitialized = True
```
both `self.is_initialized` and `self.isInitialized` exist in this branch and I assume they are meant to be the same variable. Convert `isInitialized` to `is_initialized`.

prhuft replied:
see my prior comment.

c-poole commented on hsdio.py:
```diff
@@ -63,60 +62,70 @@ def load_xml(self, node):
         'node': type is ET.Element. tag should be "HSDIO"
         """
+        
+        self.isInitialized = False
```
This is the only version to exist in this Instrument but I would still prefer the other naming convention to be applied here for the sake of consistency.

prhuft replied:
due to PXI code that checks this parameter for multiple devices, they all need to have the same name. We can globally change variables to the underscore format in a future PR.

c-poole commented on instrument.py:
```diff
@@ -206,7 +211,7 @@ def init(self):
         if not self.enable:
             return
-                    
+                                
```
Unnecessary white space

c-poole commented on instrument.py:
```diff
@@ -12,11 +12,15 @@
 import re
 
+## local class imports
+from instrumentfuncs import str_to_bool
```
str_to_bool is defined in this module instead of being imported now

c-poole commented on instrument.py:
```diff
@@ -139,6 +143,7 @@ def __init__(self, pxi, expected_root, node: ET.Element = None):
         self.pxi = pxi
         self.expectedRoot = expected_root
         self.enable = False
+        self.isInitialized = False
```
same comment for previous casing convention

prhuft replied:
same response for previous comment

c-poole commented on ni_hsdio.py:
```diff
@@ -4,9 +4,10 @@
 
 
-class HSDIOError(Exception):
+class HSDIOError(HardwareError):
```
Need to add import for HardwareError

c-poole commented on pxi.py:
```diff
@@ -7,33 +7,35 @@
 
 
-# TODO : Should this inherit from XMLLoader?
+# TODO : Should this inherit from XMLLoader? <-- i'm on board with this. - Preston
```
What utility does that provide?

prhuft replied:
Very little (logging would be set up in the super call, and the intended use of load_xml is essentially that of parse_xml). I think mostly it would just signal to readers that the spinal cord pxi is the same as an instrument. But upon second glance, XMLLoader would have to be modified slightly, and to your point, it's not worth it for now.

c-poole commented on pxi.py:
```diff
@@ -170,107 +180,104 @@ def parse_xml(self, xml_str):
+                            # get timeout in [ms]
+                            self.measurement_timeout = 1000*float(child.text)
+                        except ValueError as e:
+                            msg = f"{e} \n {child.text} is not valid "+
```
Missing \ in line continuation. Preferably use a different method like parentheses

c-poole commented on pxi.py:
```diff
@@ -170,107 +180,104 @@ def parse_xml(self, xml_str):
+                    elif child.tag == "AnalogOutput":
+                        # set up the analog_output
+                        # self.analog_output.load_xml(child)
+                        # self.analog_output.init() # setup in labview
```
Uncomment methods that have been implemented so the remaining work to be done is clear.

prhuft replied:
I had done this, then remembered you had commented out anything that would inhibit a successful dry run of the server. I will change it back.

c-poole commented on pxi.py:
```diff
@@ -293,24 +300,18 @@ def data_to_xml(self):
-        # return_data_str += self.demo.data_out()  # TODO : Implement
+        for dev in devices:
+            if dev.isInitialized: 
+                return_data_str += self.dev.data_out()
```
self.dev should just be dev here.

c-poole commented on pxi.py:
```diff
@@ -350,50 +359,97 @@ def reset_data(self):
+    device_method_call(device_list: List[Instrument], method_name: str)
+    
+    this would sacrifice some transparency but would be cleaner and lend to 
+    easier implementation of similar methods in the future. thoughts?
```
+1

prhuft replied:
:)

c-poole commented on pxi.py:
```diff
@@ -502,12 +555,75 @@ def handle_errors(self, error, traceback_str):
+        
+        # maybe there are some cases when we would want to overwrite or add to
+        # the detailed message already acquired where the exception arose?
+        if traceback_str != None:
```
Equality comparison with None

c-poole replied:
is not None

c-poole commented on pxi.py:
```diff
@@ -502,12 +555,75 @@ def handle_errors(self, error, traceback_str):
+            # TODO: log and handle timeout error
+            pass 
+            
+    def cycle_message(self, dev: XMLLoader):
```
XMLLoader has not been imported in this module

c-poole commented on analogout.py:
```diff
@@ -148,35 +159,51 @@ def init(self):
+                        try:
+                            self.task.close()
+                            
+                        except DaqError as e:
```
If we are not going to use e then there is no need to include `as e`

c-poole commented on analogout.py:
```diff
@@ -148,35 +159,51 @@ def init(self):
+                            Signal.START_TRIGGER,
+                            self.exportTrigger.outputTerminal)
+                
+                except DaqError as e:
```
same as previous. Won't mark every case like I did with self.task to reduce spam this time though.

c-poole commented on analogout.py:
```diff
@@ -148,35 +159,51 @@ def init(self):
+                            samps_per_chan=1000) # default
+                        
+                    if self.startTrigger.wait_for_start_trigger:
+                        self.task.start_trigger.cfg_dig_edge_start_trig(
```
self.startTrigger and self.task.start_trigger are both referenced here. Are the two variables used as intended or another case of our casing convention mismatch?

prhuft replied:
this is intended. startTrigger is an instance of trigger.StartTrigger, while task.start_trigger is something defined by the NIDAQmx module.

c-poole commented on digitalin.py:
```diff
@@ -56,7 +65,11 @@ def load_xml(self, node):
-                    
+                            
+                except ValueError:
+                    self.logger.exception()
```
logger.exception requires a message argument. Same comment for all other instances of this.

c-poole commented on digitalout.py:
```diff
@@ -66,16 +77,15 @@ def load_xml(self, node):
                             self.logger.warning(f"Unrecognized XML tag \'{node.tag}\' in <{child.tag}>")
                 
                 elif child.tag == "waveform":
                     self.waveform = Waveform()
```
Waveform isn't imported in this module

c-poole commented on hamamatsu.py:
```diff
@@ -111,109 +112,117 @@ def load_xml(self, node: ET.Element):
+
+                elif child.tag == "exposureTime":
+                    # can convert scientifically-formatted numbers - good
+                    self.exposure_time = self.float(child.text)
```
float is a python function not a function of this class instance

c-poole commented on hamamatsu.py:
```diff
@@ -111,109 +112,117 @@ def load_xml(self, node: ET.Element):
-            elif child.tag == "forceImagesToU16":
-                self.images_to_U16 = Instrument.str_to_bool(child.text)
+                else:
+                        self.logger.warning(f"Node {child.tag} is not a valid Hamamatsu attribute")
```
body of else block over indented

c-poole commented on hsdio.py:
```diff
@@ -63,60 +62,70 @@ def load_xml(self, node):
-            elif child.tag == "hardwareAlignmentQuantum":
-                self.hardwareAlignmentQuantum = child.text
+                elif child.tag == "clockRate":
+                    self.clockRate = self.float(child.text)
```
self.float should just be float

c-poole commented on pxierrors.py:
```diff
@@ -0,0 +1,151 @@
+
+## local class imports
+from instrument import XMLLoader
+from digitalout import *
```
why the star import? The only thing used is in the __main__ call and its one class.

c-poole commented on pxi.py:
```diff
@@ -64,10 +67,10 @@ def __init__(self, address: Tuple[str, int]):
-        # self.daqmx_do = DAQmxDO(self)
+        self.analog_input = AnalogOutput(self)
+        self.analog_output = AnalogInput(self)
+        self.ttl = TTLInput(self)
```
Need to add imports for these instruments.

c-poole commented on pxi.py:
```diff
@@ -170,107 +180,104 @@ def parse_xml(self, xml_str):
+                        
+                    elif child.tag == "DAQmxDO":
+                        self.daqmxdo.load_xml(child)
+                        self.daqmxdo.init()
```
constructor makes a self.daqmx_do

c-poole commented on pxi.py:
```diff
@@ -350,50 +359,104 @@ def reset_data(self):
+        """
+        
+        if not (self.stop_connections or self.exit_measurement):
+            for dev in devices:
```
the argument name is device_list not devices

c-poole commented on pxi.py:
```diff
@@ -350,50 +359,104 @@ def reset_data(self):
+            self.handle_errors(e)
+            
+
+    def batch_method_call(self, device_list: List[Instrument], method: str):
```
Need to import List from Typing for this type annotation

prhuft commented:
@c-poole 's changes are made, and I fixed a few other bugs I caught. The approach to error handling I've implemented is tentative, and can be changed if need be (e.g. if traceback in some places ends up being insufficient ).
